File: src/domain/jobs.ts
import type { EngineStatus, SlotId } from "./types.ts";

export type JobPhase =
  | "metadata"
  | "beat"
  | "key"
  | "stems"
  | "pitch-time"
  | "vocal-cleanup"
  | "arrangement"
  | "export";

export type JobState = "idle" | "queued" | "running" | "complete" | "failed" | "cancelled";

export interface MashJobStep {
  id: JobPhase;
  label: string;
  state: JobState;
  status: EngineStatus;
  message: string;
  details?: string[];
  startedAt: string | null;
  completedAt: string | null;
}

export interface MashTrackJob {
  jobId: string;
  sessionId: string;
  slotId: SlotId;
  inspectionId: string;
  state: JobState;
  steps: MashJobStep[];
  createdAt: string;
  updatedAt: string;
}

export const JOB_PHASE_ORDER: JobPhase[] = [
  "metadata",
  "beat",
  "key",
  "stems",
  "pitch-time",
  "vocal-cleanup",
  "arrangement",
  "export",
];

export const JOB_PHASE_LABELS: Record<JobPhase, string> = {
  metadata: "Local metadata",
  beat: "Beat / tempo / phrase",
  key: "Key / harmony",
  stems: "Stem separation",
  "pitch-time": "Pitch / time",
  "vocal-cleanup": "Vocal cleanup",
  arrangement: "Arrangement drafts",
  export: "Export / mastering",
};

export function deriveJobState(steps: MashJobStep[]): JobState {
  if (steps.some((step) => step.state === "running")) {
    return "running";
  }

  if (steps.every((step) => step.state === "idle" || step.state === "queued")) {
    return "idle";
  }

  if (steps.some((step) => step.state === "failed")) {
    return "failed";
  }

  if (steps.every((step) => step.state === "complete" || step.state === "idle")) {
    return steps.some((step) => step.state === "complete") ? "complete" : "idle";
  }

  return "running";
}

export function createTrackJob(params: {
  sessionId: string;
  slotId: SlotId;
  inspectionId: string;
}): MashTrackJob {
  const now = new Date().toISOString();

  return {
    jobId: crypto.randomUUID(),
    sessionId: params.sessionId,
    slotId: params.slotId,
    inspectionId: params.inspectionId,
    state: "idle",
    steps: JOB_PHASE_ORDER.map((phase) => ({
      id: phase,
      label: JOB_PHASE_LABELS[phase],
      state: "idle",
      status: phase === "metadata" ? "implemented" : defaultPhaseStatus(phase),
      message: defaultPhaseMessage(phase),
      startedAt: null,
      completedAt: null,
    })),
    createdAt: now,
    updatedAt: now,
  };
}

function defaultPhaseStatus(phase: JobPhase): EngineStatus {
  if (phase === "beat" || phase === "key") {
    return "analysis-coming-next";
  }

  return "engine-pending";
}

function defaultPhaseMessage(phase: JobPhase): string {
  switch (phase) {
    case "metadata":
      return "Waiting for browser metadata inspection.";
    case "beat":
      return "BeatNet+/Essentia adapter lane not connected.";
    case "key":
      return "Key detector adapter lane not connected.";
    case "stems":
      return "Demucs / HTDemucs adapter lane not connected.";
    case "pitch-time":
      return "Rubber Band adapter lane not connected.";
    case "vocal-cleanup":
      return "Vocal cleanup chain not connected.";
    case "arrangement":
      return "Arrangement draft generator not connected.";
    case "export":
      return "Export and mastering lane not connected.";
    default:
      return "Engine pending.";
  }
}

File: local-engine/service/jobs.py
"""In-memory local job store for the sidecar foundation."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from models import CreateJobRequest, JobPhase, JobResponse, JobState

_jobs: dict[str, JobResponse] = {}


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def create_job(request: CreateJobRequest) -> JobResponse:
    job_id = str(uuid4())
    timestamp = _now_iso()

    job = JobResponse(
        job_id=job_id,
        phase=request.phase,
        state="queued",
        status=_phase_status(request.phase),
        message=_phase_message(request.phase),
        session_id=request.session_id,
        slot_id=request.slot_id,
        result=request.input or None,
        created_at=timestamp,
        updated_at=timestamp,
    )
    _jobs[job_id] = job
    return job


def get_job(job_id: str) -> JobResponse | None:
    return _jobs.get(job_id)


def update_job(
    job_id: str,
    *,
    state: JobState | None = None,
    status: str | None = None,
    message: str | None = None,
    result: dict[str, Any] | None = None,
) -> JobResponse | None:
    job = _jobs.get(job_id)
    if job is None:
        return None

    updated = job.model_copy(
        update={
            "state": state or job.state,
            "status": status or job.status,
            "message": message or job.message,
            "result": result if result is not None else job.result,
            "updated_at": _now_iso(),
        }
    )
    _jobs[job_id] = updated
    return updated


def complete_metadata_job(job_id: str, metadata_result: dict[str, Any]) -> JobResponse | None:
    return update_job(
        job_id,
        state="complete",
        status="implemented",
        message="Metadata job completed through local ffprobe inspection.",
        result=metadata_result,
    )


def fail_job(job_id: str, message: str, *, status: str = "failed") -> JobResponse | None:
    return update_job(job_id, state="failed", status=status, message=message)


def _phase_status(phase: JobPhase) -> str:
    if phase == "metadata":
        return "implemented"
    if phase in {"beat", "key"}:
        return "analysis-coming-next"
    return "engine-pending"


def _phase_message(phase: JobPhase) -> str:
    messages = {
        "metadata": "Metadata job queued for local ffprobe inspection.",
        "beat": "Beat analysis is not implemented in this service phase.",
        "key": "Key analysis is not implemented in this service phase.",
        "stems": "Stem separation is not implemented in this service phase.",
        "pitch-time": "Pitch/time processing is not implemented in this service phase.",
        "vocal-cleanup": "Vocal cleanup is not implemented in this service phase.",
        "arrangement": "Arrangement drafts are not implemented in this service phase.",
        "export": "Export/mastering is not implemented in this service phase.",
    }
    return messages.get(phase, "Job queued.")
